pedsnetdcc/db.py

```diff
@@ -0,0 +1,430 @@
+        :returns:            the processed Statement object itself
+        :rtype:              Statement
+        """
+
+        local_logger = self.logger
+        conn_info = get_conn_info_dict(conn.dsn)
+
+        if logq:
+            # See the DictQueueHandler docstring for the reason why the
+            # standard logging.handlers.QueueHandler can't be used here.
+            qh = DictQueueHandler(logq)
+            local_logger = logging.getLogger()
+            local_logger.setLevel(logging.DEBUG)
+            local_logger.addHandler(qh)
+
+        try:
+            with conn.cursor() as cursor:
+
+                # Execute the query.
+                msg_dict = combine_dicts({'msg': self.msg, 'sql': self.sql,
+                                          'id_': self.id_}, conn_info)
+                local_logger.debug(msg_dict)
+
+                cursor.execute(self.sql)
+
+                # Get the effected row count.
+                if cursor.rowcount in [-1, None]:
+                    self.rowcount = None
+                else:
+                    self.rowcount = cursor.rowcount
+
+                # Get the result field names.
+                if cursor.description:
+                    for field in cursor.description:
+                        self.fields.append(field[0])
+                else:
+                    self.fields = []
+
+                # Retrieve the data or, if none, set data to None.
+                try:
+                    self.data = cursor.fetchall()
+                except psycopg2.ProgrammingError as e:
+                    if e.args[0] == 'no results to fetch':
+                        self.data = None
+                    else:
+                        raise
+
+        except Exception as err:
+            self.err = err
+            msg_dict = combine_dicts({'msg': 'database error while {0}'.
+                                      format(self.msg), 'err': str(err),
+                                      'id': self.id_}, conn_info)
+            local_logger.error(msg_dict)
+
+        else:
+            self.err = None
+
+        if resq:
+            resq.put(self)
+
+        return self
+
+
+class StatementSet(collections.MutableSet):
+    """A set of statements that can be executed in parallel.
+
+    A collections.MutableSet class that adds a `parallel_execute` method that
+    calls `execute` on each member of the set. The method intelligently handles
+    tasks for a given number of workers, logging messages and errors from
+    workers without random interleaving, and collecting the results of the work
+    back into the set.
+
+    The class is implemented with a simple underlying set in the `data`
+    attribute that can be manipulated directly if needed.
+
+    Although the intent is for all members to be Statement objects, any object
+    that meets the `obj.execute(conn_str, resq=None, logq=None) -> obj` API
+    should work (no type checking is done).
```

Comment:
Returning self allows for method chaining or immediate attribute access a la `result = obj.execute(conn_str).data`. On the other hand, it slightly obfuscates the fact that the object is being modified, as a side effect. Which is why methods like `list.sort()` do not return the list. I thought this "modify and return" strategy would give us the best of both worlds so long as we document and remember it.